src/pykeen/training/callbacks.py

```diff
@@ -374,9 +384,147 @@ def post_epoch(self, epoch: int, epoch_loss: float, **kwargs: Any) -> None:  # n
+        """
+        super().__init__()
+        self.step = step
+        self.pre_step_callbacks = tuple(pre_step_callbacks or [])
+
+    # docstr-coverage: inherited
+    def pre_batch(self, **kwargs: Any) -> None:  # noqa: D102
+        # Recall that torch *accumulates* gradients. Before passing in a
+        # new instance, you need to zero out the gradients from the old instance
+        self.optimizer.zero_grad(set_to_none=True)
+
+    # docstr-coverage: inherited
+    def post_batch(self, epoch: int, batch, **kwargs: Any) -> None:  # noqa: D102
+        # when called by batch_size_search(), the parameter update should not be applied.
+        if self.step:
+            # pre-step callbacks
+            for cb in self.pre_step_callbacks:
+                cb.pre_step(epoch=epoch, **kwargs)
+            # update parameters according to optimizer
+            self.optimizer.step()
+
+        # After changing applying the gradients to the embeddings, the model is notified that the forward
+        # constraints are no longer applied
+        # TODO: do we want that when self.step is not set?
+        self.model.post_parameter_update()
+
+
+class LearningRateSchedulerTrainingCallback(TrainingCallback):
+    """Update learning rate scheduler."""
+
+    # docstr-coverage: inherited
+    def post_epoch(self, epoch: int, epoch_loss: float, **kwargs: Any) -> None:  # noqa: D102
+        if self.training_loop.lr_scheduler is not None:
+            self.training_loop.lr_scheduler.step(epoch=epoch)
+
+
+def _hasher(kwargs: Mapping[str, Any]) -> int:
+    # do not share optimal parameters across different training loops
+    return id(kwargs["training_loop"])
+
+
+@maximize_memory_utilization(parameter_name=("batch_size", "slice_size"), hasher=_hasher)
+@torch.inference_mode()
+def _validation_loss_amo_wrapper(
+    training_loop: "pykeen.training.TrainingLoop",
+    triples_factory: CoreTriplesFactory,
+    batch_size: int,
+    slice_size: int,
+    label_smoothing: float,
+    **kwargs,
+) -> float:
+    """Calculate validation loss with automatic batch size optimization."""
+    return training_loop._train_epoch(
+        # todo: create dataset only once
+        batches=training_loop._create_training_data_loader(
+            triples_factory=triples_factory, batch_size=batch_size, drop_last=False, **kwargs
+        ),
+        label_smoothing=label_smoothing,
+        # no callbacks, no epoch
+        callbacks=MultiTrainingCallback(),
+        epoch=-1,
+        # no sub-batching (for evaluation, we can just reduce batch size without any effect)
+        sub_batch_size=None,
+        slice_size=slice_size if training_loop.supports_slicing else None,
+        # this is handled by the AMO wrapper
+        only_size_probing=False,
+        # no backward passes
+        backward=False,
+    )
+
+
+class ValidationLossTrainingCallback(TrainingCallback):
+    """Calculate loss on a development set."""
```

Comment:
can we get an end-to-end example usage in the docstring please

Reply:
2e35f1e